### ISSUE TITLE
prepare_text_and_asr_labels silently drops EOS when delay_text_channel_by shifts it into the delay window

### ISSUE THREAD

**Issue by udsy19** [NVIDIA-NeMo/NeMo#16164]
**Describe the bug**

`prepare_text_and_asr_labels`'s `delay_text_channel_by` path (`nemo/collections/speechlm2/parts/label_prep.py:162-178`) silently drops the EOS token whenever it falls within the last `delay_by` positions of `target_tokens`.

The block masks out every EOS token in that trailing window (they are about to be cut off by the right-shift that implements the delay) and tries to relocate one of them into the slot that will still be present after the shift:

```python
delay_by = cfg.get("delay_text_channel_by", 0)
if delay_by > 0:
    eos_mask = (target_tokens == text_eos_id) & (
        torch.arange(target_tokens.size(1), device=target_tokens.device).unsqueeze(0)
        >= (target_tokens.size(1) - delay_by)
    )
    for i in range(target_tokens.size(0)):
        if eos_mask[i].any():
            target_tokens[i, -(delay_by)] = text_eos_id
    target_tokens = torch.where(eos_mask, text_pad_id, target_tokens)
    pad = torch.full((target_tokens.shape[0], delay_by), fill_value=text_pad_id, ...)
    target_tokens = torch.cat([pad, target_tokens[:, :-delay_by]], dim=-1)
```

The relocation writes to `target_tokens[i, -(delay_by)]`, i.e. index `T - delay_by`. But the very next line slices with `target_tokens[:, :-delay_by]`, which keeps only indices `0 .. T - delay_by - 1` — index `T - delay_by` is exactly one of the positions this slice drops. So the "rescued" EOS is discarded along with the rest of the delay window, and the relocation never has any effect on the output. The slot that actually survives the slice-then-pad and lands at the final sequence's last position is `target_tokens[i, -(delay_by + 1)]`.

**Steps/Code to reproduce bug**

```python
import torch
from nemo.collections.speechlm2.parts.label_prep import prepare_text_and_asr_labels

PAD, BOS, EOS = 0, 1, 3
delay_by = 2
tokens = torch.tensor([[BOS, 10, 11, 12, 13, 14, EOS, PAD]])  # EOS inside the delay window

out = prepare_text_and_asr_labels(
    batch={"target_token_lens": torch.tensor([7])},
    target_tokens=tokens,
    source_encoded=torch.zeros(1, tokens.shape[1], 4),
    cfg={"delay_text_channel_by": delay_by},
    text_pad_id=PAD, text_bos_id=BOS, text_eos_id=EOS,
)
print(out["text_labels"])
```

Observed on `main` (`6281c939aa`): `tensor([[ 0,  1, 10, 11, 12, 13, 14]])` — no `3` (EOS) anywhere in the labels.
Expected: the EOS token is preserved somewhere in the output, delayed by `delay_by` steps like every other token.

An EOS well outside the delay window (e.g. at index 3 of an 8-token sequence with `delay_by=2`) survives correctly, so the loss is specific to the boundary case the rescue code targets.

**Expected behavior**

`delay_text_channel_by` should shift the whole sequence, including the EOS token, without dropping supervision for it. `target_tokens[i, -(delay_by)]` should be `target_tokens[i, -(delay_by + 1)]`.

**Environment overview**

- Environment location: Bare-metal (CPU)
- Method of NeMo install: from source (`pip install -e .`)

**Environment details**

- OS version: Ubuntu 24.04 (Linux 6.8, x86_64)
- PyTorch version: 2.13.0+cpu
- Python version: 3.12.3

**Additional context**

`prepare_text_and_asr_labels` is called from `DuplexSTTModel.prepare_inputs` (`nemo/collections/speechlm2/models/duplex_stt_model.py:252`) with the full model `cfg`, and `delay_text_channel_by` is a documented config key in the function's own docstring. The shipped `examples/speechlm2/conf/duplex_stt.yaml` does not set it, but any training run that does (a documented, supported knob) silently loses EOS supervision for every target sequence whose EOS lands in the last `delay_by` positions — a plausible occurrence for short/normal-length utterances relative to typical delay values, not just an edge case.

`git log` shows `nemo/collections/speechlm2/parts/label_prep.py` has exactly one commit since it was added (#15092, "Implement Duplex Speech-to-text model") — this branch has never been touched or exercised by a test since.
